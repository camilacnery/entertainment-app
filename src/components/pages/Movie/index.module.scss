.main {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;

  @media screen and (max-width: 1031px) {
    padding: 6rem 0 0 0;
  }

  @media screen and (min-width: 1032px) {
    padding: 4rem 0 0 0;
  }
}

.movieContainer {
  display: flex;
  margin: 2rem;
  border-radius: 5px;

  img {
    border-radius: 5px;
    align-self: center;
  }

  @media screen and (max-width: 1031px) {
    flex-direction: column;
    img {
      max-height: 48vh;
      width: auto;
    }
  }

  @media screen and (min-width: 1032px) {
    flex-direction: row;
    margin: 4rem;
    padding: 4rem;
  }
}

.movieDetails {
  @media screen and (min-width: 1032px) {
    margin-left: 2rem;
  }
}

.title {
  margin: 2rem 0 0 0;
  line-height: 1.15;
  font-size: 4rem;
  z-index: 2;

  @media screen and (max-width: 1031px) {
    font-size: 2.6rem;
  }

  @media screen and (min-width: 1032px) {
    font-size: 4rem;
  }
}

.content {
  z-index: 2;
}
