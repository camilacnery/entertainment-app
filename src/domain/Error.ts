export type ErrorStatus = 404 | 500;
