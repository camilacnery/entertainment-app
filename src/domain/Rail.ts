export interface IRailItem {
  posterUrl?: string;
  backdropUrl?: string;
  title: string;
  id: number;
}

export interface IRail {
  name: string;
  items: IRailItem[];
}
