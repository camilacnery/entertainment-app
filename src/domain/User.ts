export interface IUserProfile {
  imageUrl: string;
  fullName: string;
  firstName: string;
  email: string;
}
