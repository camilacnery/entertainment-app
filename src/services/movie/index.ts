import tmdbClient, { ITMDBMovieDetails } from "@/clients/tmdb";
import { IMovie } from "@/domain/Movie";
import { getMovieRecommendationRail } from "@/services/rails";

const getMovie = async (id: string): Promise<IMovie> => {
  try {
    const movie = await tmdbClient.request<ITMDBMovieDetails>(`/movie/${id}`);

    const recommendations = await getMovieRecommendationRail(id);
    const posterUrl = tmdbClient.buildImageUrl(movie.poster_path);
    const backdropUrl = tmdbClient.buildImageUrl(movie.backdrop_path);

    return {
      title: movie.title,
      ...(posterUrl && { posterUrl }),
      ...(backdropUrl && { backdropUrl }),
      ...(movie.original_title && { originalTitle: movie.original_title }),
      ...(movie.release_date && { releaseDate: movie.release_date }),
      ...(movie.tagline && { tagline: movie.tagline }),
      ...(movie.runtime && { runtime: movie.runtime }),
      ...(movie.overview && { description: movie.overview }),
      genres: movie.genres?.map((genre) => genre.name) || [],
      recommendations,
    };
  } catch (e) {
    console.log("[ERROR] Movie details", e);
    throw new Error("Error fetching movie details");
  }
};

export { getMovie };
