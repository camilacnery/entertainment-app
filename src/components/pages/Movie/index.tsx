import { FC } from "react";
import Head from "next/head";
import Image from "next/future/image";
import { IMovie } from "../../../domain/Movie";
import Rail from "../../general/Rail";
import Meta from "../../general/Meta";
import styles from "./index.module.scss";

type TProps = {
  movie: IMovie;
};

const MoviePage: FC<TProps> = ({ movie }) => {
  return (
    <div className={styles.container}>
      <Head>
        <Meta
          title={movie.title}
          description={movie.description || "Movie details page"}
        ></Meta>
      </Head>

      <main className={styles.main}>
        <div className={styles.banner}>
          <Image
            src={movie.backdropUrl}
            alt={`Imagem ${movie.title}`}
            width={800}
            height={600}
          />
        </div>

        <div className={styles.content}>
          <div className={styles.movie}>
            <Image
              loading="lazy"
              width={300}
              height={380}
              src={movie.posterUrl}
              alt={`Movie poster for ${movie.title}`}
            />
            <section className={styles.movieDetails}>
              <h1 className={styles.title}>{movie.title}</h1>
              <p>{`${movie.releaseDate} • ${movie.genres.join(", ")} • ${
                movie.runtime
              }min`}</p>
              <p>{movie.tagline}</p>
              <h2>Overview</h2>
              <p>{movie.description}</p>
            </section>
          </div>
          <Rail name="Similar content" items={movie.recommendations} />
        </div>
      </main>
    </div>
  );
};

export default MoviePage;
