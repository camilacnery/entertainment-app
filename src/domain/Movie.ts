import { TRailItem } from "./Rail";

export type TMovie = {
  posterUrl: string;
  backdropUrl: string;
  title: string;
  originalTitle: string;
  releaseDate: string;
  tagline: string;
  description: string;
  genres: string[];
  recommendations: TRailItem[];
};
