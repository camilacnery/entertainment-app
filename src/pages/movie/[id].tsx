import type { GetStaticProps, NextPage } from "next";
import MoviePage from "@/components/pages/Movie";
import { IMovie } from "@/domain/Movie";
import { getMovie } from "@/services/movie";
import { ErrorStatus } from "@/domain/Error";
import ErrorPage from "@/components/pages/Error";

const Movie: NextPage<{ movie: IMovie; errorStatus: ErrorStatus }> = (props) => {
  if (props.errorStatus) return <ErrorPage status={props.errorStatus} />;

  return <MoviePage {...props} />;
};

export const getStaticProps: GetStaticProps = async (context) => {
  try {
    const id = (context?.params?.id as string) || "";
    const movie = await getMovie(id);

    console.log("aquii", movie);
    return {
      props: { movie },
    };
  } catch {
    return {
      props: { errorStatus: 500, revalidate: 60 },
    };
  }
};

export const getStaticPaths = () => {
  return {
    paths: [],
    fallback: "blocking",
  };
};

export default Movie;
