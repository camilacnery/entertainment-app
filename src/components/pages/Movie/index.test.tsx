import { IMovie } from "@/domain/Movie";
import { buildMovie } from "@/tests/builders/movie";
import { screen, render, within } from "@testing-library/react";

import Movie from ".";

const pageElements = {
  rail: {
    list: (name: string) => screen.queryByRole("list", { name }),
    items: (rail: HTMLElement) => within(rail).getAllByRole("listitem"),
  },
  movieDetails: {
    title: (name: string) => screen.getByRole("heading", { name }),
    releaseDate: (date: string) => screen.getByText(new RegExp(date)),
    genres: (genres: string) => screen.getByText(new RegExp(genres)),
    duration: (duration: string) => screen.getByText(new RegExp(duration)),
    tagline: (tagline: string) => screen.getByText(tagline),
    description: (description: string) => screen.getByText(description),
  },
};

const renderMovie = (movie: IMovie) => {
  return render(<Movie movie={movie} />);
};

describe("Movie", () => {
  it("renders movie details with all content", () => {
    const movie = buildMovie();
    const { baseElement } = renderMovie(movie);

    const recommendations = pageElements.rail.list("Similar movies")!;

    expect(baseElement).toMatchSnapshot();

    expect(pageElements.movieDetails.title("Movie Title")).toBeInTheDocument();
    expect(pageElements.movieDetails.releaseDate("2022-09-16")).toBeInTheDocument();
    expect(pageElements.movieDetails.genres("Action, Drama")).toBeInTheDocument();
    expect(pageElements.movieDetails.duration("167min")).toBeInTheDocument();
    expect(pageElements.movieDetails.tagline("Movie tagline")).toBeInTheDocument();
    expect(pageElements.movieDetails.tagline("Movie description")).toBeInTheDocument();
    expect(pageElements.rail.items(recommendations)).toHaveLength(4);
  });
});
