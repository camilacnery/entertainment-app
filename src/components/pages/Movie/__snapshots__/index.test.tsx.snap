// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie renders movie details with all content 1`] = `
<body>
  <div>
    <div
      class="banner"
    >
      <img
        alt=""
        data-nimg="future"
        decoding="async"
        height="768"
        src="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FkXfqcdQKsToO0OUXHcrrNCHDBzO.jpg&w=2048&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FkXfqcdQKsToO0OUXHcrrNCHDBzO.jpg&w=1080&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FkXfqcdQKsToO0OUXHcrrNCHDBzO.jpg&w=2048&q=75 2x"
        style="color: transparent;"
        width="1024"
      />
    </div>
    <main
      class="main"
    >
      <div
        class="content"
      >
        <div
          class="content"
        >
          <img
            alt=""
            data-nimg="future"
            decoding="async"
            height="380"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2Fq6y0Go1tsGEsmtFryDOJo3dEmqu.jpg&w=640&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2Fq6y0Go1tsGEsmtFryDOJo3dEmqu.jpg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2Fq6y0Go1tsGEsmtFryDOJo3dEmqu.jpg&w=640&q=75 2x"
            style="color: transparent;"
            width="300"
          />
          <section
            class="movieDetails"
          >
            <h1
              class="title"
            >
              The Shawshank Redemption
            </h1>
            <p>
              1994-09-23 • Drama, Crime • 142min
            </p>
            <p>
              Fear can hold you prisoner. Hope can set you free.
            </p>
            <h2>
              Overview
            </h2>
            <p>
              Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.
            </p>
          </section>
        </div>
        <section
          class="rail"
        >
          <h1
            id="rail-heading-Recommendations"
          >
            Recommendations
          </h1>
          <ul
            aria-labelledby="rail-heading-Recommendations"
            class="railList"
          >
            <li
              class="railListItem"
            >
              <span
                class="railListItemTitle"
              >
                Schindler's List
              </span>
              <img
                alt="Schindler's List"
                data-nimg="future"
                decoding="async"
                height="300"
                loading="lazy"
                src="http://localhost/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FsF1U4EUQS8YHUYjNl3pMGNIQyr0.jpg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FsF1U4EUQS8YHUYjNl3pMGNIQyr0.jpg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FsF1U4EUQS8YHUYjNl3pMGNIQyr0.jpg&w=640&q=75 2x"
                style="color: transparent;"
                width="200"
              />
            </li>
            <li
              class="railListItem"
            >
              <span
                class="railListItemTitle"
              >
                The Godfather
              </span>
              <img
                alt="The Godfather"
                data-nimg="future"
                decoding="async"
                height="300"
                loading="lazy"
                src="http://localhost/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2F3bhkrj58Vtu7enYsRolD1fZdja1.jpg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2F3bhkrj58Vtu7enYsRolD1fZdja1.jpg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2F3bhkrj58Vtu7enYsRolD1fZdja1.jpg&w=640&q=75 2x"
                style="color: transparent;"
                width="200"
              />
            </li>
            <li
              class="railListItem"
            >
              <span
                class="railListItemTitle"
              >
                Pulp Fiction
              </span>
              <img
                alt="Pulp Fiction"
                data-nimg="future"
                decoding="async"
                height="300"
                loading="lazy"
                src="http://localhost/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FfIE3lAGcZDV1G6XM5KmuWnNsPp1.jpg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FfIE3lAGcZDV1G6XM5KmuWnNsPp1.jpg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FfIE3lAGcZDV1G6XM5KmuWnNsPp1.jpg&w=640&q=75 2x"
                style="color: transparent;"
                width="200"
              />
            </li>
            <li
              class="railListItem"
            >
              <span
                class="railListItemTitle"
              >
                Forrest Gump
              </span>
              <img
                alt="Forrest Gump"
                data-nimg="future"
                decoding="async"
                height="300"
                loading="lazy"
                src="http://localhost/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FsaHP97rTPS5eLmrLQEcANmKrsFl.jpg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FsaHP97rTPS5eLmrLQEcANmKrsFl.jpg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FsaHP97rTPS5eLmrLQEcANmKrsFl.jpg&w=640&q=75 2x"
                style="color: transparent;"
                width="200"
              />
            </li>
            <li
              class="railListItem"
            >
              <span
                class="railListItemTitle"
              >
                Dilwale Dulhania Le Jayenge
              </span>
              <img
                alt="Dilwale Dulhania Le Jayenge"
                data-nimg="future"
                decoding="async"
                height="300"
                loading="lazy"
                src="http://localhost/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2F2CAL2433ZeIihfX1Hb2139CX0pW.jpg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2F2CAL2433ZeIihfX1Hb2139CX0pW.jpg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2F2CAL2433ZeIihfX1Hb2139CX0pW.jpg&w=640&q=75 2x"
                style="color: transparent;"
                width="200"
              />
            </li>
          </ul>
        </section>
      </div>
    </main>
  </div>
</body>
`;
